From e0692f378db37db17c153eb7d1dd18a094a8a532 Mon Sep 17 00:00:00 2001
From: Eli Friedman <efriedma@quicinc.com>
Date: Tue, 3 Jun 2025 09:51:37 -0700
Subject: [PATCH] [clang] Don't evaluate the initializer of constexpr-unknown
 parameters. (#142498)

If we see a parameter of reference type that isn't part of the frame,
don't try to evaluate its default argument. Just treat it as a
constexpr-unknown value.

Fixes #141114.  Fixes #141858.
---
 clang/lib/AST/ExprConstant.cpp                |   7 +-
 .../SemaCXX/constant-expression-p2280r4.cpp   | 152 ++++++++++--------
 2 files changed, 88 insertions(+), 71 deletions(-)

diff --git a/clang/lib/AST/ExprConstant.cpp b/clang/lib/AST/ExprConstant.cpp
index b79d8c197fe7..4b73c8f836d0 100644
--- a/clang/lib/AST/ExprConstant.cpp
+++ b/clang/lib/AST/ExprConstant.cpp
@@ -3524,7 +3524,12 @@ static bool evaluateVarDeclInit(EvalInfo &Info, const Expr *E,
   // should begin within the evaluation of E
   // Used to be C++20 [expr.const]p5.12.2:
   // ... its lifetime began within the evaluation of E;
-  if (isa<ParmVarDecl>(VD) && !AllowConstexprUnknown) {
+  if (isa<ParmVarDecl>(VD)) {
+    if (AllowConstexprUnknown) {
+      Result = &Info.CurrentCall->createConstexprUnknownAPValues(VD, Base);
+      return true;
+    }
+
     // Assume parameters of a potential constant expression are usable in
     // constant expressions.
     if (!Info.checkingPotentialConstantExpression() ||
diff --git a/clang/test/SemaCXX/constant-expression-p2280r4.cpp b/clang/test/SemaCXX/constant-expression-p2280r4.cpp
index 0cdc16ed4e82..61e96cbb2af4 100644
--- a/clang/test/SemaCXX/constant-expression-p2280r4.cpp
+++ b/clang/test/SemaCXX/constant-expression-p2280r4.cpp
@@ -5,46 +5,46 @@ using size_t = decltype(sizeof(0));
 
 namespace std {
 struct type_info {
-  const char* name() const noexcept(true);
+const char* name() const noexcept(true);
 };
 }
 
 template <typename T, size_t N>
 constexpr size_t array_size(T (&)[N]) {
-  return N;
+return N;
 }
 
 void use_array(int const (&gold_medal_mel)[2]) {
-  constexpr auto gold = array_size(gold_medal_mel); // ok
+constexpr auto gold = array_size(gold_medal_mel); // ok
 }
 
 constexpr auto olympic_mile() {
-  const int ledecky = 1500;
-  return []{ return ledecky; };
+const int ledecky = 1500;
+return []{ return ledecky; };
 }
 static_assert(olympic_mile()() == 1500); // ok
 
 struct Swim {
-  constexpr int phelps() { return 28; }
-  virtual constexpr int lochte() { return 12; }
-  int coughlin = 12;
+constexpr int phelps() { return 28; }
+virtual constexpr int lochte() { return 12; }
+int coughlin = 12;
 };
 
 constexpr int how_many(Swim& swam) {
-  Swim* p = &swam;
-  return (p + 1 - 1)->phelps();
+Swim* p = &swam;
+return (p + 1 - 1)->phelps();
 }
 
 void splash(Swim& swam) {
-  static_assert(swam.phelps() == 28);     // ok
-  static_assert((&swam)->phelps() == 28); // ok
-  Swim* pswam = &swam;                    // expected-note {{declared here}}
-  static_assert(pswam->phelps() == 28);   // expected-error {{static assertion expression is not an integral constant expression}} \
-                                          // expected-note {{read of non-constexpr variable 'pswam' is not allowed in a constant expression}}
-  static_assert(how_many(swam) == 28);    // ok
-  static_assert(Swim().lochte() == 12);   // ok
-  static_assert(swam.lochte() == 12);     // expected-error {{static assertion expression is not an integral constant expression}}
-  static_assert(swam.coughlin == 12);     // expected-error {{static assertion expression is not an integral constant expression}}
+static_assert(swam.phelps() == 28);     // ok
+static_assert((&swam)->phelps() == 28); // ok
+Swim* pswam = &swam;                    // expected-note {{declared here}}
+static_assert(pswam->phelps() == 28);   // expected-error {{static assertion expression is not an integral constant expression}} \
+				  // expected-note {{read of non-constexpr variable 'pswam' is not allowed in a constant expression}}
+static_assert(how_many(swam) == 28);    // ok
+static_assert(Swim().lochte() == 12);   // ok
+static_assert(swam.lochte() == 12);     // expected-error {{static assertion expression is not an integral constant expression}}
+static_assert(swam.coughlin == 12);     // expected-error {{static assertion expression is not an integral constant expression}}
 }
 
 extern Swim dc;
@@ -52,129 +52,141 @@ extern Swim& trident; // interpreter-note {{declared here}}
 
 constexpr auto& sandeno   = typeid(dc);         // ok: can only be typeid(Swim)
 constexpr auto& gallagher = typeid(trident);    // expected-error {{constexpr variable 'gallagher' must be initialized by a constant expression}} \
-                                                // interpreter-note {{initializer of 'trident' is unknown}}
+					// interpreter-note {{initializer of 'trident' is unknown}}
 
 namespace explicitThis {
 struct C {
-  constexpr int b()  { return 0; };
+constexpr int b()  { return 0; };
 
-  constexpr int f(this C &c) {
-    return c.b();     // ok
-  }
+constexpr int f(this C &c) {
+return c.b();     // ok
+}
 
-   constexpr int g() {
-    return f();       // ok
-  }
+constexpr int g() {
+return f();       // ok
+}
 };
 
 void g() {
-  C c;
-  constexpr int x = c.f();
-  constexpr int y = c.g();
+C c;
+constexpr int x = c.f();
+constexpr int y = c.g();
 }
 }
 
 namespace GH64376 {
 template<int V>
 struct Test {
-    static constexpr int value = V;
+static constexpr int value = V;
 };
 
 int main() {
-    Test<124> test;
-    auto& test2 = test;
+Test<124> test;
+auto& test2 = test;
 
-    if constexpr(test2.value > 3) {
-       return 1;
-    }
+if constexpr(test2.value > 3) {
+return 1;
+}
 
-    return 0;
+return 0;
 }
 }
 
 namespace GH30060 {
 template<int V>
 struct A {
-  static constexpr int value = V;
+static constexpr int value = V;
 };
 
 template<class T>
 static void test1(T &f) {
-    A<f.value> bar;
+A<f.value> bar;
 }
 
 void g() {
-    A<42> f;
+A<42> f;
 
-    test1(f);
+test1(f);
 }
 }
 
 namespace GH26067 {
 struct A {
-    constexpr operator int() const { return 42; }
+constexpr operator int() const { return 42; }
 };
 
 template <int>
 void f() {}
 
 void test(const A& value) {
-    f<value>();
+f<value>();
 }
 
 int main() {
-    A a{};
-    test(a);
+A a{};
+test(a);
 }
 }
 
 namespace GH34365 {
 void g() {
-  auto f = []() { return 42; };
-  constexpr int x = f();
-  [](auto f) { constexpr int x = f(); }(f);
-  [](auto &f) { constexpr int x = f(); }(f);
-  (void)[&]() { constexpr int x = f(); };
+auto f = []() { return 42; };
+constexpr int x = f();
+[](auto f) { constexpr int x = f(); }(f);
+[](auto &f) { constexpr int x = f(); }(f);
+(void)[&]() { constexpr int x = f(); };
 }
 }
 
 namespace GH118063 {
 template <unsigned int N>
 struct array {
-    constexpr auto size() const -> unsigned int {
-        return N;
-    }
+constexpr auto size() const -> unsigned int {
+return N;
+}
 };
 
 constexpr auto f(array<5> const& arr) {
-    return array<arr.size()>{}.size();
+return array<arr.size()>{}.size();
 }
 
 int g() {
-    array<5> arr {};
-    static_assert(f(arr) == 5);
+array<5> arr {};
+static_assert(f(arr) == 5);
 }
 }
 
 namespace GH128409 {
-  int &ff();
-  int &x = ff(); // expected-note {{declared here}}
-  constinit int &z = x; // expected-error {{variable does not have a constant initializer}} \
-                        // expected-note {{required by 'constinit' specifier here}} \
-                        // expected-note {{initializer of 'x' is not a constant expression}}
+int &ff();
+int &x = ff(); // expected-note {{declared here}}
+constinit int &z = x; // expected-error {{variable does not have a constant initializer}} \
+		// expected-note {{required by 'constinit' specifier here}} \
+		// expected-note {{initializer of 'x' is not a constant expression}}
 }
 
 namespace GH129845 {
-  int &ff();
-  int &x = ff(); // expected-note {{declared here}}
-  struct A { int& x; };
-  constexpr A g = {x}; // expected-error {{constexpr variable 'g' must be initialized by a constant expression}} \
-                       // expected-note {{initializer of 'x' is not a constant expression}}
-  const A* gg = &g;
+int &ff();
+int &x = ff(); // expected-note {{declared here}}
+struct A { int& x; };
+constexpr A g = {x}; // expected-error {{constexpr variable 'g' must be initialized by a constant expression}} \
+	       // expected-note {{initializer of 'x' is not a constant expression}}
+const A* gg = &g;
 }
 
 namespace extern_reference_used_as_unknown {
-  extern int &x;
-  int y;
-  constinit int& g = (x,y); // expected-warning {{left operand of comma operator has no effect}}
+extern int &x;
+int y;
+constinit int& g = (x,y); // expected-warning {{left operand of comma operator has no effect}}
+}
+
+namespace param_reference {
+  constexpr int arbitrary = -12345;
+  constexpr void f(const int &x = arbitrary) { // expected-note {{declared here}}
+    constexpr const int &v1 = x; // expected-error {{must be initialized by a constant expression}} \
+    // expected-note {{reference to 'x' is not a constant expression}}
+    constexpr const int &v2 = (x, arbitrary); // expected-warning {{left operand of comma operator has no effect}}
+    constexpr int v3 = x; // expected-error {{must be initialized by a constant expression}}
+    static_assert(x==arbitrary); // expected-error {{static assertion expression is not an integral constant expression}}
+    static_assert(&x - &x == 0);
+  }
 }
-- 
2.53.0.1018.g2bb0e51243-goog

